### PR TITLE
Enable using the Ninja generator instead of Visual Studio on Windows builds.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
## Summary for Infra Rollout

We're enabling users to use Ninja across the coreclr, libraries, and host native builds instead of Visual Studio or Make. Ninja is a slimmed down build orchestrator that performs much faster than MSBuild on our native Windows CoreCLR build and performs comparably to MSBuild on the Windows libraries and host builds and Make on all non-Windows builds.

You can use the `-ninja` flag to use Ninja to run the native builds instead of MSBuild or Make. For this rollout, we're not changing the default experience for local development, but we are enabling Ninja builds in CI for the CoreCLR build.

There is one breaking change and one non-breaking change for developers:

- Breaking: If you are using the `src/coreclr/build-runtime.cmd` script to build on Windows, you'll need to build either the `clr.dactools` or `clr.runtime` subset once from the root script first. The `DacTableGen` tool used in the Windows build has moved out of the CMake build and is now a regular C# project. As a result, it needs to be built via the root build script once so it is available for the native build.
- Non breaking: The minimum recommended CMake version is 3.16. If you use CMake 3.16 precompiled headers will be used in all builds. If you use CMake 3.15.5 or lower, precompiled headers will not be used. This is a behavior change, since we used to always have precompiled headers on Windows and never have them on non-Windows. If you notice slowness in your Windows build after pulling down this PR and have not updated CMake, try installing CMake 3.16 to re-enable precompiled headers.

## Original Summary

This PR enables using the Ninja generator on Windows instead of the Visual Studio CMake generators via the `-ninja` switch (for the build-runtime.cmd script) and via a `Ninja` MSBuild global property. To keep pre-compiled headers working on Windows, this PR bumps the minimum recommended CMake version to 3.16. Using an older CMake version is still possible, but your build will not support precompiled headers. However, now precompiled headers are enabled on all platforms (not just Windows) when using CMake 3.16 or higher.

Since the C# support in CMake is only for the Visual Studio generators, I've moved the DacTableGen project into a subset that's included with the clr.runtime subset and also moved the PIA for DIALib into a checked-in ilproj project by decompiling it with ildasm.

To support opening the CoreCLR VS-generated solution in Visual Studio, I've augmented the root `-vs` flag. If you pass `coreclr.sln` (case-insensitive) to the `-vs` flag, the build script will generate the CoreCLR solution for the provided parameters using the VS generator (if it hasn't been generated already) and open the solution in VS.

I've updated the CoreCLR, Libraries, and Host builds to all enable using Ninja.

For this PR, Ninja usage will be opt-in. In the future, we will make Ninja the default generator and make the Visual Studio generator opt-in.

TODO:

- [x] CI validation
	- The CI machines don't have Ninja installed yet, so this is blocked on getting it installed.
- [x] Documentation of new options

### CODE REVIEW

**Comment by hoyosjs** [comment]
> Since the C# support in CMake is only for the Visual Studio generators, I've moved the DacTableGen project into a subset that's included with the clr.runtime subset and also moved the PIA for DIALib into a checked-in ilproj project by decompiling it with ildasm.

A lot of people use the `build-runtime.cmd`. I am definitely not a fan of managed projects in the native script, but we have prior art (opt, version headers, etc), is there any reason this one falls out from there? As for the `ilproj`, are we still doing restore, or are we using the live version of the IL SDK?

File: src/coreclr/build-runtime.cmd [hunk 1]
```diff
@@ -86,6 +86,7 @@ set __RestoreOptData=1
 set __CrossArch=
 set __PgoOptDataPath=
 set __CMakeArgs=
+set __Ninja=0
```
Comment (jashook) [inline src/coreclr/build-runtime.cmd:89]: I think I would suggest setting ninja as the default build target.

File: eng/native/configurecompiler.cmake [hunk 2]
```diff
@@ -425,26 +425,41 @@ endif(CLR_CMAKE_HOST_UNIX)
...
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/WX>) # treat warnings as errors
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/Oi>) # enable intrinsics
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/Oy->) # disable suppressing of the creation of frame pointers on the call stack for quicker function calls
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/Gm->) # disable minimal rebuild
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/Zp8>) # pack structs on 8-byte boundary
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/Gy>) # separate functions for linker
+
+  # disable C++ RTTI
+  # /GR is added by default by CMake, so remove it manually.
+  string(REPLACE "/GR" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
```
Comment (janvorli) [inline eng/native/configurecompiler.cmake:440]: It seems that we can use `add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/GR>)` instead. 
Also, we don't need to remove the `/GR` from the flags. The Visual C++ doc says that if CL encounters conflicting options, it uses the rightmost one. I think that the ones that cmake defines as default could be viewed as overridable defaults.

File: eng/native/configurecompiler.cmake [hunk 3]
```diff
@@ -464,22 +484,30 @@ if (MSVC)
...
 
   if (CLR_CMAKE_HOST_ARCH_I386)
-    add_compile_options(/Gz)
+    add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/Gz>)
   endif (CLR_CMAKE_HOST_ARCH_I386)
 
   add_compile_options($<$<OR:$<CONFIG:Release>,$<CONFIG:Relwithdebinfo>>:/GL>)
-  add_compile_options($<$<OR:$<OR:$<CONFIG:Release>,$<CONFIG:Relwithdebinfo>>,$<CONFIG:Checked>>:/O1>)
+  add_compile_options($<$<CONFIG:Checked>:/O1>)
+  add_compile_options($<$<OR:$<CONFIG:Release>,$<CONFIG:Relwithdebinfo>>:/Ob2iy->)
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:510]: This change is preserving the current optimization settings. I'll do it in a separate PR to validate it.

File: src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt [hunk 4]
```diff
@@ -127,6 +127,7 @@ if(CLR_CMAKE_TARGET_WIN32)
         shlwapi.lib
         bcrypt.lib
         RuntimeObject.lib
+        delayimp.lib
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt:130]: This was included by default by MSBuild even with setting the default libraries list to empty.

File: src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt [hunk 5]
```diff
@@ -180,51 +181,45 @@ if(FEATURE_SINGLE_FILE_DIAGNOSTICS)
 endif(FEATURE_SINGLE_FILE_DIAGNOSTICS)
 
 if(CLR_CMAKE_TARGET_WIN32)
-    if (NOT CMAKE_GENERATOR MATCHES "Visual Studio .*")
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt:183]: We can remove this since we aren't building DacTableGen with the C# language support in CMake (that's the part that was Visual Studio only).

Additionally, we're removing the NMake support and replacing it with Ninja. To my knowledge, the only usage of NMake was to generate a compile_commands.json file for jit-format, and Ninja can do that as well.

File: eng/native/configurecompiler.cmake (hunk 2 above)
Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:440]: CL does use the rightmost option, but it emits an unsuppressable warning when doing so.

**Comment by jashook** [comment]
@jkoritzinsky can you post the binary sizes before and after this change?

**Comment by jkoritzinsky** [comment]
@jashook I'll pull those together tomorrow.

The binary size differences we were seeing were because the prototype didn't pass the configuration to cmake at config time, so we were always building as debug. Once I updated the scripts to pass the configuration correctly, release builds were approximately the same size.

**Comment by jashook** [comment]
Interesting I thought I was comparing checked to checked. Which I guess would have a difference between checked and debug. Do you mind posting numbers for all three configs?

Thank you!

**Comment by jkoritzinsky** [comment]
I’ll pull together sizes for each config and share them.

**Comment by jkoritzinsky** [comment]
I do see a regression in Checked. It looks like the Ninja build is building Checked as though it is Debug. I'll go fix it.

File: eng/native/functions.cmake [hunk 6]
```diff
@@ -331,7 +285,16 @@ function(strip_symbols targetName outputFilename)
 
     set(${outputFilename} ${strip_destination_file} PARENT_SCOPE)
   else(CLR_CMAKE_HOST_UNIX)
-    set(${outputFilename} ${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/${targetName}.pdb PARENT_SCOPE)
+    get_property(_isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+    if(_isMultiConfig)
+      # We can't use the $<TARGET_PDB_FILE> generator expression here since
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/functions.cmake:290]: Yes, when I first wrote this, we end up coming down this path for the resource dlls. It's possible with the numerous refactorings since then that this has changed.

File: src/coreclr/src/debug/di/CMakeLists.txt [hunk 7]
```diff
@@ -79,5 +79,5 @@ endif (CLR_CMAKE_TARGET_WIN32)
 list(APPEND CORDBDI_SOURCES ${CORDBDI_SOURCES_ASM_FILE})
 
 add_library_clr(cordbdi STATIC ${CORDBDI_SOURCES})
-target_precompile_header(TARGET cordbdi HEADER stdafx.h)
+target_precompile_headers(cordbdi PRIVATE stdafx.h)
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/src/debug/di/CMakeLists.txt:82]: The `[["stdafx.h"]]` syntax adds the literal include "stdafx.h", whereas the `stdafx.h` syntax adds an include for the stdafx.h file in the directory. Since we don't always use `CMAKE_INCLUDE_SOURCE_DIR`, I've used the `stdafx.h` syntax over the literal syntax where possible.

File: src/coreclr/src/dlls/dbgshim/CMakeLists.txt [hunk 8]
```diff
@@ -9,7 +9,7 @@ set(DBGSHIM_SOURCES
 
 if(CLR_CMAKE_HOST_WIN32)
     # Use static crt
-    add_definitions(-MT)
+    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:/MT>)
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/src/dlls/dbgshim/CMakeLists.txt:12]: I think this is a typo. I'll fix this in a future commit.

File: eng/native/configurecompiler.cmake [hunk 9]
```diff
@@ -425,26 +425,41 @@ endif(CLR_CMAKE_HOST_UNIX)
...
-  add_compile_options(/Zp8) # pack structs on 8-byte boundary
-  add_compile_options(/Gy) # separate functions for linker
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /GR-") # disable C++ RTTI
-  add_compile_options(/FC) # use full pathnames in diagnostics
-  add_compile_options(/MP) # Build with Multiple Processes (number of processes equal to the number of processors)
-  add_compile_options(/Zm200) # Specify Precompiled Header Memory Allocation Limit of 150MB
-  add_compile_options(/Zc:strictStrings) # Disable string-literal to char* or wchar_t* conversion
-
-  add_compile_options(/wd4960 /wd4961 /wd4603 /wd4627 /wd4838 /wd4456 /wd4457 /wd4458 /wd4459 /wd4091 /we4640)
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/TP>) # compile all files as C++
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:428]: These changes were needed because both MASM and RC are CMake langauges, so `add_compile_options` was adding all of these flags to files of those languages as well.

File: eng/native/configurecompiler.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake]: I've removed this particular change since we were already passing the various flags at different times.

File: src/coreclr/build-runtime.cmd (hunk 1 above)
Comment (jkoritzinsky) [reply src/coreclr/build-runtime.cmd:89]: For this PR, I'm planning on making Ninja opt-in. That way we can merge this in outside of an infra rollout. In a future infra rollout, we'll make Ninja the default.

File: eng/native/functions.cmake (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/functions.cmake:266]: Yep we still call `strip_symbols` for resource dlls.

**Comment by jkoritzinsky** [comment]
@jashook Here's the current numbers for the size of coreclr.dll in master and in the Ninja build in this branch:

| CoreCLR size (bytes) | MSBuild build | Ninja build |
|----------------------|---------------|-------------|
| Debug                | 31730688      | 31341568    |
| Checked              | 13066752      | 13064704    |
| Release              | 5145088       | 4938240     |

Ninja build sizes are comparable if not better.

**Comment by trylek** [comment]
@jkoritzinsky - I guess I was naively assuming that Ninja is basically just a different orchestration mechanism and the underlying compilation and linking processes basically remain unchanged - if that is the case, what is the reason for any size diffs at all? I can imagine that the 2KB size diff in Checked mode is truly negligible and can be due to minute details like object file ordering and / or some alignment hackery in the linker or some such; but I'm having a hard time to find a benign explanation of the 200 KB difference in Release mode and 400 KB difference in Debug mode.

I'm wondering whether we might want to try to compare some of the compiler command lines, object file dumps and / or the linker maps to get a better understanding of the diff. I mean, it's kind of cool that Ninja produces a slightly smaller binary but if it's at the cost of embedded debug info in Debug mode or if that means that we're passing around completely different sets of compiler optimization flags in the Release mode, it might turn out to be a less pleasant surprise

**Comment by jkoritzinsky** [comment]
The difference is likely because MSBuild has a lot of built-in assumptions that pass various flags that override flags that we were passing in, in particular the optimization flags. In this PR, I've tried to make the flags more explicit.

I think it would be worth comparing the flags or and the outputs to figure out where the savings came from. (My guess is the optimization flags)

**Comment by janvorli** [comment]
> I think it would be worth comparing the flags or and the outputs to figure out where the savings came from. (My guess is the optimization flags)

We should do that to make sure we are not missing something.

**Comment by jkoritzinsky** [comment]
I've started by comparing the size outputs of the MSBuild build on this branch to master. I can't seem to find any difference in the Debug build project files outside of the PCH changes, but the MSBuild output on this branch is still a different size than master.

For release, there looks to be a change from MinSpace optimization to MaxSpeed, so that's something we can change back if so desired.

**Comment by trylek** [comment]
Back in .NET Native I once wrote a comparator of linker (rhbind at that time) map files to produce structured size diffs we used to chase down various size regressions. If you're able to make the linker produce a pair of map files and share them, I can try to take a look if I might be able to adapt my work from that time to create a simple C# structured comparator tool; if it turns out to be easy, it should be a piece of cake to run it on all binaries we're interested in and see where the differences are.

**Comment by jkoritzinsky** [comment]
I'll work on generating the linker map files from master and from the ninja build in this branch.

**Comment by jkoritzinsky** [comment]
@trylek I have the builds generating map files locally. Let me know once we've got some way to compare them to track down the change.

**Comment by trylek** [comment]
@jkoritzinsky - Could you please share one or two sample pairs, either as GitHub gists, by sending them to me as e-mail attachments or by uploading them to the team share under \\clrmain\public\writable\users\jkoritzinsky? I'll experiment with them over the weekend.

**Comment by jkoritzinsky** [comment]
I've uploaded all of the linker map files for a debug build to `\\clrmain\public\writable\users\jekoritz\ninja-linker-map-files`. Hopefully that will have what you need.

File: eng/native/configurecompiler.cmake (hunk 9 above)
[the remark this answers is not in the document]
Comment (janvorli) [reply eng/native/configurecompiler.cmake:436]: Actually, multiple languages in the COMPILE_LANGUAGE are not supported until CMake 3.15, according to cmake doc. Some of the test legs in this CI are failing due to that. Also, the doc discourages using this generator expression:
> This generator expression has limited use because it is not possible to use it with the Visual Studio generators. Portable buildsystems would not use this expression, and would create separate libraries for each source file language instead.

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:436]: Can you point me to where that doc is from? I haven’t seen that.

[the remark this answers is not in the document]
Comment (janvorli) [reply eng/native/configurecompiler.cmake:436]: It is in the COMPILE_LANGUAGE doc: https://cmake.org/cmake/help/v3.6/manual/cmake-generator-expressions.7.html?highlight=compile_language. The doc for 3.15+ use a little different wording, but it is there as well.

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/configurecompiler.cmake:436]: The docs for 3.15+ describe the behavior I expect, with the VS generator treating everything as C++. Since all cases we use this generator expressions here we apply to both C and C++, we’re in the supported zone.

Re outdated CMake on the CI machines: I’m working on getting 3.16 installed on CI along with Ninja so the new precompiled header functionality works on CI machines.

**Comment by jkoritzinsky** [comment]
This PR is ready for review!

File: src/libraries/Native/build-native.proj [hunk 10]
```diff
@@ -15,7 +16,7 @@
 
   <Target Name="BuildNativeUnix"
           BeforeTargets="Build"
-          Condition="'$(TargetOS)' != 'Windows_NT'">
+          Condition="!$([MSBuild]::IsOsPlatform(Windows))">
```
Comment (ViktorHofer) [inline src/libraries/Native/build-native.proj:19]: Unsure if this ever worked but the TargetOS check doesn't assert the current OS but the target one that is passed in via the `-os` switch.

Comment (ViktorHofer) [reply src/libraries/Native/build-native.proj:19]: Can I build the native bits targeting Unix on Windows?

Comment (jkoritzinsky) [reply src/libraries/Native/build-native.proj:19]: This broke with the WASM on Windows build, since it was going down the `build-native.sh` path when it should have gone down the `build-native.cmd` path.

File: src/coreclr/src/debug/daccess/dacfn.cpp [hunk 11]
```diff
@@ -104,6 +104,8 @@ DacWarning(__in char* format, ...)
     OutputDebugStringA(text);
 }
 
+#pragma warning (push)
+#pragma warning (disable: 4297)
```
Comment (janvorli) [inline src/coreclr/src/debug/daccess/dacfn.cpp:108]: I wonder why this was not needed before and I wonder if throwing the exception from the functions below will actually work as expected. My worry is that the compiler might exclude EH info for these.

File: src/libraries/Native/build-native.cmd [hunk 12]
```diff
@@ -93,6 +97,7 @@ goto :SetupDirs
 echo Commencing build of native components
 echo.
 
+if /i "%__BuildArch%" == "wasm" set __sourceDir=%~dp0..\Unix
```
Comment (janvorli) [inline src/libraries/Native/build-native.cmd:100]: It is not clear to me how is this change related to the build using ninja

File: src/coreclr/runtime.proj [hunk 13]
```diff
@@ -20,7 +20,7 @@
                                    '$(NoPgoOptimize)' != 'true'"
                                    Include="-enforcepgo" />
       <_CoreClrBuildArg Condition="$([MSBuild]::IsOsPlatform(Windows)) and '$(CrossDac)' != ''" Include="-$(CrossDac)dac" />
-      <_CoreClrBuildArg Condition="'$(NoPgoOptimize)' == 'true'" Include="-nopgooptimize" />
+      <_CoreClrBuildArg Condition="'$(Ninja)' == 'true'" Include="-ninja" />
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/runtime.proj:23]: That option should still be included below. I'll make sure I didn't delete it.

File: src/libraries/Native/build-native.cmd (hunk 12 above)
Comment (jkoritzinsky) [reply src/libraries/Native/build-native.cmd:100]: The refactoring of the `gen-buildsys` scripts so that we use a common one for Windows moved this line out of the libraries version of the `gen-buildsys` script and up to `build-native.cmd`. I refactored the scripts so that we have a common entry point for generating the CMake project for CoreCLR, Libraries, and Host (same as on Linux) since that makes handling determining the CMake parameters for the generator easier.

File: src/coreclr/build-runtime.cmd [hunk 14]
```diff
@@ -532,8 +500,52 @@ if %__BuildCrossArchNative% EQU 1 (
...
+
+        set __CMakeBinDir="%__CrossComponent2BinDir%"
+        set "__CMakeBinDir=!__CMakeBinDir:\=/!"
+        
+        if %__Ninja% EQU 1 (
+            set __ExtraCmakeArgs="-DCMAKE_BUILD_TYPE=!__BuildType!" 
+        )
+
+        set __ExtraCmakeArgs=!__ExtraCmakeArgs! %__CMakeClrBuildSubsetArgs% "-DCLR_CROSS_COMPONENTS_BUILD=1" "-DCLR_CMAKE_TARGET_ARCH=%__BuildArch%" "-DCLR_CMAKE_TARGET_OS=%__TargetOS%" "-DCLR_CMAKE_PGO_INSTRUMENT=0" "-DCLR_CMAKE_OPTDATA_PATH=%__PgoOptDataPath%" "-DCLR_CMAKE_PGO_OPTIMIZE=0" "-DCMAKE_SYSTEM_VERSION=10.0" "-DCLR_ENG_NATIVE_DIR=%__RepoRootDir%/eng/native" "-DCLR_REPO_ROOT_DIR=%__RepoRootDir%" %__CMakeArgs%
+        call "%__RepoRootDir%\eng\native\gen-buildsys.cmd" "%__ProjectDir%" "%__CrossComp2IntermediatesDir%" %__VSVersion% %__CrossArch2% !__ExtraCmakeArgs!
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/build-runtime.cmd:531]: This echo was for debugging. I meant to remove it.

File: eng/native/functions.cmake [hunk 15]
```diff
@@ -389,8 +364,11 @@ function(install_clr)
           endif()
 
           if(CLR_CMAKE_PGO_INSTRUMENT)
-              if(WIN32)
-                  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/${targetName}.pgd DESTINATION ${destination}/PGD OPTIONAL)
+              get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+              if(is_multi_config AND WIN32)
+                install(FILES ${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/${targetName}.pgd DESTINATION ${destination}/PGD OPTIONAL)
+              else()
+                install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${targetName}.pgd DESTINATION ${destination}/PGD OPTIONAL)
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/functions.cmake:371]: How does this change behavior on non-Windows? We've always used the non-`$<CONFIG>` path on non-Windows and will continue to do so in all cases. The only change is on Windows where we use the non-`$<CONFIG>` path when we aren't using a multi-config generator (which is required for this to work with Ninja).

[the remark this answers is not in the document]
Comment (janvorli) [reply eng/native/functions.cmake:371]: We were not performing this install on non-windows before your change. It was just 
``` cmake
              if(WIN32)
                  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/${targetName}.pgd DESTINATION ${destination}/PGD OPTIONAL)
              endif()
```

[the remark this answers is not in the document]
Comment (janvorli) [reply eng/native/functions.cmake:371]: So I guess this change should rather look like this:
```cmake 
if(WIN32)
    if(is_multi_config)
        install(FILES ${CMAKE_CURRENT_BINARY_DIR}/$<CONFIG>/${targetName}.pgd DESTINATION ${destination}/PGD OPTIONAL)
    else()
        install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${targetName}.pgd DESTINATION ${destination}/PGD OPTIONAL)
    endif()
endif()
```

File: src/coreclr/src/debug/daccess/dacfn.cpp (hunk 11 above)
Comment (jkoritzinsky) [reply src/coreclr/src/debug/daccess/dacfn.cpp:108]: I just tested this out again locally by removing this pragma and it didn't emit the warning (I added this when I picked this up again since it had started warning randomly). I'll remove the pragmas so we can make sure to not miss the warning if it pops again.

File: src/coreclr/runtime.proj (hunk 13 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/runtime.proj:23]: I've validated locally that the NoPgoOptimize setting just moved to line 28.

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/coreclr/runtime.proj:23]: We just had a duplicate/it looks like a duplicate after some merges and re-arranging.

File: eng/native/functions.cmake (hunk 15 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply eng/native/functions.cmake:371]: You are correct, I misread the code. I'll change it to not install on non-Windows.

**Review by janvorli — APPROVED** [review]

**Comment by BruceForstall** [comment]
Looks like you need to add more documentation on Ninja, e.g.:
1. From where do I install it?
2. How do I install it?
3. What is the minimum version?

The answers to these questions need to cover all platforms.

The documents at the various "Requirements" links on https://github.com/dotnet/runtime/blob/master/docs/workflow/README.md should be updated.

**Comment by BruceForstall** [comment]
> The minimum recommended CMake version is 3.16

I didn't look thoroughly: is there an easily visible warning on the console if you don't have the recommended minimum cmake version? (i.e., it seems like there should be)

**Comment by jkoritzinsky** [comment]
We have an error for the minimum required CMake, but not a warning for the minimum recommended CMake (this is the first time they differ). I'll add a warning output.

File: docs/workflow/requirements/windows-requirements.md [hunk 16]
```diff
@@ -42,7 +42,13 @@ The dotnet/runtime repository requires at least Visual Studio 2019 16.6.
 - Add its location (e.g. C:\Program Files (x86)\CMake\bin) to the PATH environment variable.
   The installation script has a check box to do this, but you can do it yourself after the fact following the instructions at [Adding to the Default PATH variable](#adding-to-the-default-path-variable).
 
-The dotnet/runtime repository requires at least CMake 3.15.5.
+The dotnet/runtime repository recommends using CMake 3.16.0 or newer, but works with CMake 3.15.5.
+
+## Ninja (optional)
+
+- Install Ninja in one of the two following ways
+  - [Download the executable](https://github.com/ninja-build/ninja/releases) and add it's location to [the Default PATH variable](#adding-to-the-default-path-variable).
```
Comment (BruceForstall) [inline docs/workflow/requirements/windows-requirements.md:50]: nit: it's => its

Comment (jkoritzinsky) [reply docs/workflow/requirements/windows-requirements.md:50]:
```suggestion
  - [Download the executable](https://github.com/ninja-build/ninja/releases) and add its location to [the Default PATH variable](#adding-to-the-default-path-variable).
```
